From ccf6f5551277a4dd0dfb6828409e7022bac0f435 Mon Sep 17 00:00:00 2001
From: Alberto Fustinoni <alberto@aftnet.net>
Date: Mon, 28 Sep 2020 22:36:31 +0900
Subject: [PATCH 2/2] sokol_app changes for android support

---
 sokol_app.h | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/sokol_app.h b/sokol_app.h
index 6ad2c1a..c10759b 100644
--- a/sokol_app.h
+++ b/sokol_app.h
@@ -1622,6 +1622,7 @@ typedef struct {
     EGLDisplay display;
     EGLContext context;
     EGLSurface surface;
+    bool (*gamepad_event_handler)(const AInputEvent*);
 } _sapp_android_t;
 
 #endif // _SAPP_ANDROID
@@ -7024,6 +7025,11 @@ _SOKOL_PRIVATE int _sapp_android_input_cb(int fd, int events, void* data) {
             continue;
         }
         int32_t handled = 0;
+        if (_sapp.android.gamepad_event_handler) {
+            if (_sapp.android.gamepad_event_handler(event)) {
+                handled = 1;
+            }
+        }
         if (_sapp_android_touch_event(event) || _sapp_android_key_event(event)) {
             handled = 1;
         }
-- 
2.37.3

