From 665e5b796b00defe915c0c91ba51de0fc5ecbc99 Mon Sep 17 00:00:00 2001
From: Jon Mayo <jon@rm-f.net>
Date: Wed, 21 Sep 2022 07:15:05 -0700
Subject: [PATCH] fix initializer in gamepad API

---
 sokol_gamepad.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/sokol_gamepad.h b/sokol_gamepad.h
index 145b40c..53d2d4c 100644
--- a/sokol_gamepad.h
+++ b/sokol_gamepad.h
@@ -239,7 +239,7 @@ typedef struct sgamepad {
 #endif
 } sgamepad;
 
-_SOKOL_PRIVATE sgamepad _sgamepad = {0};
+_SOKOL_PRIVATE sgamepad _sgamepad;
 
 /*== Windows Implementation ============================================*/
 #if defined (_WIN32)
-- 
2.37.3

